File: lang/javascript/example/src/orderProcessor.js
"use strict";

const { SFNClient, StartExecutionCommand, DescribeExecutionCommand } = require("@aws-sdk/client-sfn");

/**
 * Start an order processing Step Functions execution and return the result.
 *
 * Accepts an optional pre-configured SFNClient. In production, omit the client
 * and it will be created with default settings. In tests, pass
 * `session.client("stepfunctions")` to point at the local-web-services emulator.
 *
 * @param {string} orderId
 * @param {string} stateMachineArn
 * @param {import("@aws-sdk/client-sfn").SFNClient} [sfnClient]
 * @returns {Promise<object>}
 */
async function processOrder(orderId, stateMachineArn, sfnClient) {
  const sfn = sfnClient ?? new SFNClient({ region: "us-east-1" });

  const { executionArn } = await sfn.send(
    new StartExecutionCommand({
      stateMachineArn,
      input: JSON.stringify({ orderId }),
    })
  );

  return pollUntilComplete(sfn, executionArn);
}

async function pollUntilComplete(sfn, executionArn) {
  while (true) {
    const result = await sfn.send(new DescribeExecutionCommand({ executionArn }));
    if (result.status === "SUCCEEDED") {
      return JSON.parse(result.output);
    }
    if (["FAILED", "TIMED_OUT", "ABORTED"].includes(result.status)) {
      throw new Error(`Execution ended with status: ${result.status}`);
    }
    await new Promise((r) => setTimeout(r, 100));
  }
}

module.exports = { processOrder };

File: lang/python/example/src/order_processor.py
import boto3
import json
import time


def process_order(order_id: str, state_machine_arn: str, sfn_client=None) -> dict:
    """
    Start an order processing Step Functions execution and return the result.

    Accepts an optional pre-configured boto3 SFN client. In production, omit
    the client and one will be created with default settings. In tests, pass
    ``session.client("stepfunctions")`` to point at the local-web-services emulator.
    """
    sfn = sfn_client or boto3.client("stepfunctions", region_name="us-east-1")
    response = sfn.start_execution(
        stateMachineArn=state_machine_arn,
        input=json.dumps({"orderId": order_id}),
    )
    return _poll_until_complete(sfn, response["executionArn"])


def _poll_until_complete(sfn, execution_arn: str) -> dict:
    while True:
        result = sfn.describe_execution(executionArn=execution_arn)
        status = result["status"]
        if status == "SUCCEEDED":
            return json.loads(result["output"])
        if status in ("FAILED", "TIMED_OUT", "ABORTED"):
            raise RuntimeError(f"Execution ended with status: {status}")
        time.sleep(0.1)
